reward/rewardConfigCache.go

```diff
@@ -0,0 +1,156 @@
+	InvalidFormat        = errors.New("invalid format")
+	ParsingError         = errors.New("parsing error")
+)
+
+const (
+	maxRewardConfigCache = 5 // Must be positive integer
+)
+
+type rewardConfig struct {
+	blockNum uint64
+
+	mintingAmount *big.Int
+	cnRatio       *big.Int
+	pocRatio      *big.Int
+	kirRatio      *big.Int
+	totalRatio    *big.Int
+	unitPrice     *big.Int
+}
+
+// Cache for parsed reward parameters from governance
+type rewardConfigCache struct {
+	cache            *lru.ARCCache
+	governanceHelper governanceHelper
+}
+
+func newRewardConfigCache(governanceHelper governanceHelper) *rewardConfigCache {
+	cache, _ := lru.NewARC(maxRewardConfigCache)
+	return &rewardConfigCache{
+		cache:            cache,
+		governanceHelper: governanceHelper,
+	}
+}
+
+func (rewardConfigCache *rewardConfigCache) get(blockNumber uint64) (*rewardConfig, error) {
+	var epoch uint64
+	if result, err := rewardConfigCache.governanceHelper.GetItemAtNumberByIntKey(blockNumber, params.Epoch); err == nil {
+		epoch = result.(uint64)
+	} else {
+		logger.Error("Couldn't get epoch from governance", "blockNumber", blockNumber, "err", err)
+		return nil, FailGettingConfigure
+	}
+
+	if blockNumber%epoch == 0 {
+		blockNumber -= epoch
+	} else {
+		blockNumber -= (blockNumber % epoch)
+	}
+
+	if rewardConfigCache.cache.Contains(blockNumber) {
+		config, _ := rewardConfigCache.cache.Get(blockNumber)
+		return config.(*rewardConfig), nil
+	}
+
+	newConfig, err := rewardConfigCache.newRewardConfig(blockNumber)
+	if err != nil {
+		return nil, err
+	}
+
+	rewardConfigCache.add(blockNumber, newConfig)
+	return newConfig, nil
+}
+
+func (rewardConfigCache *rewardConfigCache) newRewardConfig(blockNumber uint64) (*rewardConfig, error) {
+	mintingAmount := big.NewInt(0)
+	if result, err := rewardConfigCache.governanceHelper.GetItemAtNumberByIntKey(blockNumber, params.MintingAmount); err == nil {
+		mintingAmount.SetString(result.(string), 10)
+	} else {
+		logger.Error("Couldn't get MintingAmount from governance", "blockNumber", blockNumber, "err", err)
+		return nil, FailGettingConfigure
+	}
+
+	cnRatio := big.NewInt(0)
+	pocRatio := big.NewInt(0)
+	kirRatio := big.NewInt(0)
+	totalRatio := big.NewInt(0)
+	if result, err := rewardConfigCache.governanceHelper.GetItemAtNumberByIntKey(blockNumber, params.Ratio); err == nil {
+		cn, poc, kir, parsingError := rewardConfigCache.parseRewardRatio(result.(string))
+		if parsingError != nil {
+			return nil, parsingError
+		}
+		cnRatio.SetInt64(int64(cn))
+		pocRatio.SetInt64(int64(poc))
+		kirRatio.SetInt64(int64(kir))
+		totalRatio.SetInt64(int64(cn + poc + kir))
+	} else {
+		logger.Error("Couldn't get Ratio from governance", "blockNumber", blockNumber, "err", err)
+		return nil, FailGettingConfigure
+	}
+
+	unitPrice := big.NewInt(0)
+	if result, err := rewardConfigCache.governanceHelper.GetItemAtNumberByIntKey(blockNumber, params.UnitPrice); err == nil {
```

Comment:
How about processing err case only in this if statement?
In that case, we don't need else case.

Reply:
reflected